accord-core/src/test/java/accord/utils/SimpleBitSetTest.java

```diff
@@ -0,0 +1,203 @@
+            assertEquals(canonCollect, testCollect);
+        }
+
+        void forIndices(Random random, IntConsumer consumer)
+        {
+            for (int c = 0 ; c < 100 ; ++c)
+            {
+                int i = random.nextInt(size + 1);
+                consumer.accept(i);
+            }
+        }
+
+        void forRanges(Random random, BiConsumer<Integer, Integer> consumer)
+        {
+            for (int c = 0 ; c < 100 ; ++c)
+            {
+                int i = random.nextInt(size + 1);
+                int j = random.nextInt(size + 1);
+                if (i > j) { int t = i; i = j; j = t; }
+                consumer.accept(i, j);
+            }
+        }
+
+        static Check generate(Random random, int maxSize, int modCount, int runLength, float runChance, float clearChance)
+        {
+            int size = random.nextInt(maxSize);
+            runLength = Math.min(size, runLength);
+            BitSet canon = new BitSet(size);
+            SimpleBitSet test = new SimpleBitSet(size);
+            if (size > 0)
+            {
+                while (modCount-- > 0)
+                {
+                    boolean set = random.nextFloat() >= clearChance;
+                    boolean run = runLength > 1 && random.nextFloat() < runChance;
+                    if (run)
+                    {
+                        int i = random.nextInt(size);
+                        int j = random.nextInt(size);
+                        if (j < i) { int t = i; i = j; j = t; }
+                        j = Math.min(i + 1 + random.nextInt(runLength - 1), j);
+
+                        if (set)
+                        {
+                            canon.set(i, j);
+                            test.setRange(i, j);
+                        }
+                        else
+                        {
+                            canon.clear(i, j);
+                            while (i < j)
+                                test.unset(i++);
+                        }
+                    }
+                    else
+                    {
+                        int i = random.nextInt(size);
+                        if (set)
+                        {
+                            assertEquals(!canon.get(i), test.set(i));
+                            canon.set(i);
+                        }
+                        else
+                        {
+                            assertEquals(canon.get(i), test.unset(i));
+                            canon.clear(i);
+                        }
+                    }
+                    assertEquals(canon.cardinality(), test.setBitCount());
+                }
+            }
+            return new Check(test, canon, size);
+        }
+    }
+
+    @Test
+    public void testRandomBitSets()
+    {
+        Random random = new Random();
+        long seed = random.nextLong();
+        System.err.println("Seed: " + seed);
+        random.setSeed(seed);
+        testRandomBitSets(random, 100000);
```

Comment:
I'm really not a huge fan of the qt stuffs, but I guess this is pretty innocuous.

Reply:
`qt` is our own function / logic.  The reason I recommend this type of stuff rather than rolling our own random logic is that when failures are found most hand rolled solutions don't offer ways to repeat or even to know the seed when run in CI... by using `qt` you only have to think about your core logic and not how this runs in CI.

Reply:
my feedback branch switches

Reply:
So this logic is a very common pattern in the code already, and already supports trivially running a single example at any point by passing the relevant sub seed, which is the work of maybe 20s to establish the first time you debug. It’s a common pattern because it’s how I’ve written randomised tests in the project since time immemorial. I prefer this pattern, I am not a fan of orchestration layers, and I will not likely migrate to using qt in future. But I anm happy to accept this change in this case since you ask.

Reply:
>  and already supports trivially running a single example at any point by passing the relevant sub seed, which is the work of maybe 20s to establish the first time you debug

And when it runs in CI you don't know the seed it ran with... so you can do the code change to run it, but you don't know the seed...  logging is very problematic in CI as we *hope* that the logs got captured in JUnit and we *link* to the correct test (in this case there is only one test), and that they were not truncated...  This leads to working some times, and failing other times... its a common bug with this pattern.

> I prefer this pattern, I am not a fan of orchestration layers, and I will not likely migrate to using qt in future

This pattern is something I normally have to fix when I see failures in CI as we loose all knowledge... had to fix this in Simulator, BurnTest, Maelstorm's SimpleRandomTest, and other places in C*....  It is fine not to reuse, but when a hand roll one the setup it really must think about what happens when failures are detected in CI, and I keep finding this is lacking (which makes it very hard to retry locally)... we can keep hand rolling, but every one I have seen has done it incorrectly and I tend to fix later... so I prefer reusing when we can so we don't think about these things again

Reply:
What exactly is the issue you have found? Because I’ve never had any issue finding the failure and reproducing, but perhaps CI has gotten worse since I last did this. Either way, that sounds like a general logging problem we will have with other tests, that we want to address, a rather than an argument in favour of a favoured toolchain.

Reply:
> What exactly is the issue you have found? 

logging a seed is problematic for a few reasons:

1) if the class has multiple tests, the logs may not line up properly in the JUnit XML.  It is common to see 0 or multiple seeds in a single test class as the logs do not properly lineup with the correct method.  You may also see a single seed, but it's for a different test case.  
2) JUnit can and will truncate logs if there is too much, so when we log before the test we may be impacted by this truncation.
3) The class may finish before the logging flushes, which means the seed gets lost
4) Jenkins does not keep the log output and 100% relies on the JUnit XML file.  Circle CI keeps it, so for classes with a single test you can rely on the log file rather than the XML file, but this now requires developers to find the log file for the failing test (you look at the test report, find the class, go to the archive, find the split it ran in, then find the test class log file, then download it); Circle CI is going away, so we will be moving forward with the Jenkins behavior and rely only on the JUnit XML...

The common solution for this is to use a top level exception that includes the seed

```
try
{
  // do work
}
catch (Throwable t)
{
    throw new AssertionError("Error detected for seed " + seed, t);
}
```

JUnit will see the exception, so we know this will always make it into the JUnit XML; we don't worry about logging.  This is my main solution for each example I have given above...

Now, once we fix it so we throw an exception, we can take the seed and replay it, but if you want to add a break point *before* the failing case, you need to add a break point at the point of code that throws an issue, then go backwards to find the loop, then find the loop variable (such as `I = 109584`, then create a new break point at the loop where you stop right before the test...  this is annoying... in `qt` builder we have ability to denote that the test is *pure* or not, which means *pure* tests (such as this one) report the seed used for that single attempt, so when it gets replayed the failing test is the first iteration (rather than the 109584 in my example).  This is a usability improvement and not a "correctness" issue.

> a rather than an argument in favour of a favoured toolchain.

you can also look at this as a code duplication thing, rather than having the same code over and over again, you just have a single function

```
fuzz().run(YourClass::doWork);
```

rather than having to do

```
long seed = yourSeedGeneratorHere();
RandomSource rand = new RandomSource(seed);
int I = 0;
try
{
    for (; I < 10_000; I++)
      doWork(rand);
}
catch (Throwable t)
{
    throw new AssertionError("Error for seed " + seed + " at iteration " + i, t);
}
```

I don't personally see value in hand rolling as it's more verbose.  Now, there are cases where it totally makes sense to roll your own (such as Simulator), but for a property style test, do we really need a new framework?


NOTE:
* I use `qt` function to make it easier to migrate off QuickTheories once accord lands in trunk... My goal is to remove that dependency and just change the imports to point to the accord version...  we can name it w/e we want, but that is the history behind the name

Reply:
To be clear, I am not a fan either of quick theories itself _or_ the way in which the current abstraction on which it is based encourages tests to be written. My preferred approach is to create some class that pairs some `test` with some `canon`(ical) version, and to randomise the generation and verification of these within the class. There should then ideally be some meta-randomisation around the parameters to each of these phases, particularly the generation phase.

I think the `qt` abstraction encourages relatively narrow tests, which does permit for a quicker broader array of unique tests to be written, but the tests themselves tend to each be comparatively narrow, so there is less multiplexing of behaviour. It also seems to make it less natural to express meta-randomisation, though in principle it might be possible to do that _for_ you, this is something that should really be expressed by the test author and any API we encourage should make that a core feature/requirement.

I would be in favour of producing some better tools for meta-randomisation, where we want to declare not only the range of values but the ways in which they are randomised (rather than simply uniform over a statically defined range). I'm not sure exactly what that would look like, but this would be the most valuable tooling for improving fuzz testing IMO.

Reply:
> I strongly feel duplication is a waste and a drag on the code

What duplication are we discussing here? Because most of this is micro boiler plate. I am relatively neutral on avoiding modest amounts of boilerplate. Everything comes with a cost, including de-duplication, so unless it is a significant saving of duplication I don't see _significant_ value.

> this is what qt does...

Could you show me how? Because I do not see it doing this. It very much looks to me like you ask it for uniformly generated values. Now, you _can_ layer your own distributions on top of this, as you can with any source of randomness and Turing complete language. But I do not see any facilities provided by the QT style, and today it would make your declarations much more complex and harder to read, to the point where the API would probably be counter-productive IMO.

> This 100% depends on the author; you can write very narrow tests, or very complex ones... I do not see what in the API makes this harder than hand rolling?

APIs encourage and discourage certain behaviours. Declarative interfaces like the `qt` one are a bit like streams - they make it feel very easy to write certain kinds of simple things, but other things become much harder - particularly in comparison. So it discourages things that are too complex to express in this declarative way. Of course, you _can_ do it, but it is no longer very nice to read. Since the simpler tests are pleasing to look at, and more easily arrived at with the API, this tends to be what gets produced.

In particular, past a certain point you have to switch back to imperative programming, and I personally find layering complex imperative work _inside_ declarative frameworks makes code much harder to understand.

> for this ticket you can solve this problem by throws

I have for now already adopted your use of `qt()`. Although I would perhaps prefer to have a separate `Fuzzing.run(count, LongConsumer)` that is separate from the `qt()` tooling.